lib/format_erf.c

```diff
@@ -645,31 +645,154 @@ static int erf_read_packet(libtrace_t *libtrace, libtrace_packet_t *packet) {
-                erfptr->rlen = htons(caplen + framinglen);
+        } while(1);
+
+        return true;
+}
+
+int libtrace_to_erf_hdr(libtrace_out_t *libtrace, libtrace_packet_t *packet,
+    dag_record_t *erf, int *framinglen, int *caplen, int *padding) {
+
+    int wlen = trace_get_wire_length(packet);
+    *framinglen = dag_record_size + erf_get_padding(packet);
+    *caplen = trace_get_capture_length(packet);
+    *padding = 0;
+
+    if (*caplen <= 0 || *caplen > 65536) {
+        trace_set_err_out(libtrace, TRACE_ERR_BAD_PACKET,
+            "Capture length is out of range in libtrace_to_erf_hdr()");
+        return -1;
+    }
+
+    if (*framinglen > 65536) {
+        trace_set_err_out(libtrace, TRACE_ERR_BAD_PACKET,
+            "Framing length is to large in libtrace_to_erf_hdr()");
+        return -1;
+    }
+
+    if (*caplen + *framinglen <= 0 || *caplen + *framinglen > 65536) {
+        trace_set_err_out(libtrace, TRACE_ERR_BAD_PACKET,
+            "Capture + framing length is out of range in libtrace_to_erf_hdr()");
+        return -1;
+    }
+
+    erf->ts = bswap_host_to_le64(trace_get_erf_timestamp(packet));
+    memset(&erf->flags, 0, sizeof(erf->flags));
+    if (trace_get_direction(packet) != TRACE_DIR_UNKNOWN)
+        erf->flags.iface = trace_get_direction(packet);
+    if (!find_compatible_linktype(libtrace,packet))
+        return -1;
+    erf->type = libtrace_to_erf_type(trace_get_link_type(packet));
+
+    // calculate correct padding for the ERF record
+    erf_calculate_padding(packet, framinglen, caplen, padding);
+
+    erf->rlen = htons(*framinglen + *caplen + *padding);
+    // loss counter. Cant't do this
+    erf->lctr = 0;
+    erf->wlen = htons(wlen);
+
+    return 0;
+}
+
+static inline int erf_increased_padding(int framinglen, int caplen) {
+    return (sizeof(uint64_t) - ((framinglen + caplen) % sizeof(uint64_t)))
+        % sizeof(uint64_t);
+}
+
+static inline int erf_reduced_padding(int framinglen, int caplen) {
+    return (framinglen + caplen) % sizeof(uint64_t);
+}
 
-	if ((numbytes = 
-		wandio_wwrite(OUTPUT->file, 
-				erfptr,
-				(size_t)(framinglen))) 
-			!= (int)(framinglen)) {
+int erf_calculate_padding(libtrace_packet_t *packet, int *framinglen,
+    int *caplen, int *padding) {
+
+    int wlen = trace_get_wire_length(packet);
+    *caplen = trace_get_capture_length(packet);
+    *padding = 0;
+
+    if (packet->type == TRACE_RT_DATA_ERF)
+        *framinglen = trace_get_framing_length(packet);
+    else
+        *framinglen = dag_record_size + erf_get_padding(packet);
+
+    // full packet, round up
+    if ((wlen == *caplen) ||
+        // full packet except FCS, round up. Do we want to do this? or should
+        // we round down??
+        (trace_get_link_type(packet) == TRACE_TYPE_ETH && wlen-4 == *caplen)) {
+
+        *padding = erf_increased_padding(*framinglen, *caplen);
+    // snapped packet
+    } else {
+        *caplen -= erf_reduced_padding(*framinglen, *caplen);
```

Comment:
I'm not sure we ever want to do this? We should never reduce the cap length to align a packet only ever increase the padding.
It is common to want to capture only the first X number of bytes of a packet because that is the header you care about. So cutting this short is unexpected. Or am I missing the purpose of this code here?

Reply:
resolved in 4c690d6